### PR TITLE
CSPL-1206: Introducing the App Source scope clusterWithPreConfig

### PR DESCRIPTION
**Pull request by sgontla**
Support for new App Source scope `clusterWithPreConfig`

### CODE REVIEW

**Comment by sgontla** [comment]
With the changes listed here in this PR, first-time ES app installation works fine. However, when the Deployer Pod resets for any reason(say, for example, a new app added or existing app modified), then it is getting into crashLoopBackoff. 

When the new deployer Pod is coming up, Ansible playbooks validate all the installed apps. In case if the ES app is available, it runs the post set up playbooks, which is failing with the following error
```
TASK [splunk_deployer : Run ESS post-install setup] ****************************
fatal: [localhost]: FAILED! => {
    "changed": false,
    "cmd": [
        "/opt/splunk/bin/splunk",
        "search",
        "| essinstall --ssl_enablement ignore --deployment_type shc_deployer",
        "-auth",
        "admin:2YM3giXfE2XS0BaYWpcHwu5V"
    ],
    "delta": "0:00:00.583384",
    "end": "2021-07-27 16:55:38.110817",
    "failed_when_result": true,
    "rc": 1,
    "start": "2021-07-27 16:55:37.527433"
}

STDERR:

Unknown search command 'essinstall'.


MSG:

non-zero return code
```
Had a discussion with Jeff and Julien, and we all believe the command eesinstall is not recognized anymore, as the app was already installed, configured, and deployed to the SHC members. (Ideally, this should be the same issue even with the traditional way of deploying the ES app), and is outside the scope of the Operator. 

To work around this behavior on the Ansible side added the following fix, and it works fine, for all the Pod reset scenarios. This is just to make sure that the ES installation works in the big picture of the Operator framework, and not the final fix at this time, as we may miss the genuine error cases.
 
```
sgontla-MBP-8CAF6:splunk-ansible sgontla$ git diff roles/splunk_common/tasks/premium_apps/configure_ess.yml
diff --git a/roles/splunk_common/tasks/premium_apps/configure_ess.yml b/roles/splunk_common/tasks/premium_apps/configure_ess.yml
index 5bddbac..d978c46 100644
--- a/roles/splunk_common/tasks/premium_apps/configure_ess.yml
+++ b/roles/splunk_common/tasks/premium_apps/configure_ess.yml
@@ -19,6 +19,6 @@
   become_user: "{{ splunk.user }}"
   register: essinstall
   changed_when: essinstall.rc == 0
-  failed_when: essinstall.rc != 0
+  failed_when: essinstall.rc != 0 and 'Unknown search command' not in essinstall.stderr
   notify:
     - Restart the splunkd service
sgontla-MBP-8CAF6:splunk-ansible sgontla$
```

File: pkg/splunk/enterprise/configuration.go [hunk 1]
```diff
@@ -929,17 +929,15 @@ func validateSplunkAppSources(appFramework *enterpriseApi.AppFrameworkSpec, loca
...
 
 		if appSrc.Scope != "" {
-			if localScope && appSrc.Scope != "local" {
+			if localScope && appSrc.Scope != enterpriseApi.ScopeLocal {
 				return fmt.Errorf("Invalid scope for App Source: %s. Only local scope is supported for this kind of CR", appSrc.Name)
 			}
 
-			if appSrc.Scope != "local" && appSrc.Scope != "cluster" {
+			if !(appSrc.Scope == enterpriseApi.ScopeLocal || appSrc.Scope == enterpriseApi.ScopeCluster || appSrc.Scope == enterpriseApi.ScopeClusterWithPreConfig) {
 				return fmt.Errorf("Scope for App Source: %s should be either local or cluster", appSrc.Name)
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration.go:937]: Probably modify the error msg too, something like - "Scope for App Source: %s should be either local or cluster or clusterWithPreConfig"

File: pkg/splunk/enterprise/configuration.go [hunk 2]
```diff
@@ -949,11 +947,11 @@ func validateSplunkAppSources(appFramework *enterpriseApi.AppFrameworkSpec, loca
 
 	}
 
-	if localScope && appFramework.Defaults.Scope != "" && appFramework.Defaults.Scope != "local" {
+	if localScope && appFramework.Defaults.Scope != "" && appFramework.Defaults.Scope != enterpriseApi.ScopeLocal {
 		return fmt.Errorf("Invalid scope for defaults config. Only local scope is supported for this kind of CR")
 	}
 
-	if appFramework.Defaults.Scope != "" && appFramework.Defaults.Scope != "local" && appFramework.Defaults.Scope != "cluster" {
+	if appFramework.Defaults.Scope != "" && appFramework.Defaults.Scope != enterpriseApi.ScopeLocal && appFramework.Defaults.Scope != enterpriseApi.ScopeCluster {
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration.go:954]: Should we also add `enterpriseApi.ScopeClusterWithPreConfig` in the if condition?

File: pkg/splunk/controller/util.go [hunk 3]
```diff
@@ -183,6 +183,18 @@ func MergePodSpecUpdates(current *corev1.PodSpec, revised *corev1.PodSpec, name
...
+			if current.Containers[idx].LivenessProbe != nil && revised.Containers[idx].LivenessProbe != nil &&
+				current.Containers[idx].LivenessProbe.InitialDelaySeconds != revised.Containers[idx].LivenessProbe.InitialDelaySeconds {
+				current.Containers[idx].LivenessProbe.InitialDelaySeconds = revised.Containers[idx].LivenessProbe.InitialDelaySeconds
+				result = true
+			}
+
+			if current.Containers[idx].ReadinessProbe != nil && revised.Containers[idx].ReadinessProbe != nil &&
+				current.Containers[idx].ReadinessProbe.InitialDelaySeconds != revised.Containers[idx].ReadinessProbe.InitialDelaySeconds {
+				current.Containers[idx].ReadinessProbe.InitialDelaySeconds = revised.Containers[idx].ReadinessProbe.InitialDelaySeconds
+				result = true
```
Comment (sgontla) [inline pkg/splunk/controller/util.go:196]: cover it in UT

Comment (sgontla) [reply pkg/splunk/controller/util.go:196]: addresed

File: pkg/splunk/enterprise/configuration.go [hunk 4]
```diff
@@ -717,8 +718,11 @@ func updateSplunkPodTemplateWithConfig(client splcommon.ControllerClient, podTem
 	}
 
 	// Add extraEnv from the CommonSplunkSpec config to the extraEnv variable list
-	for _, envVar := range spec.ExtraEnv {
-		extraEnv = append(extraEnv, envVar)
+	// Exclude MC as it derives the Spec from multiple CRs
```
Comment (smohan-splunk) [inline pkg/splunk/enterprise/configuration.go:721]: Add TODO for MC CR cc: @kashok-splunk

File: pkg/splunk/enterprise/configuration.go [hunk 5]
```diff
@@ -738,12 +742,19 @@ func updateSplunkPodTemplateWithConfig(client splcommon.ControllerClient, podTem
...
+func getLivenessProbe(cr splcommon.MetaObject, instanceType InstanceType, spec *enterpriseApi.CommonSplunkSpec, additionalDelay int32) *corev1.Probe {
 	scopedLog := log.WithName("getLivenessProbe").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
 
-	livenessDelay := livenessProbeDefaultDelaySec + additionalDelay
-	if spec.LivenessInitialDelaySeconds > livenessDelay {
-		livenessDelay = spec.LivenessInitialDelaySeconds
+	livenessDelay := int32(livenessProbeDefaultDelaySec)
+
+	// Exclude MC, as it derives the spec from Multiple CRs.
+	if instanceType != SplunkMonitoringConsole {
```
Comment (smohan-splunk) [inline pkg/splunk/enterprise/configuration.go:751]: Add TODO for MC CR cc: @kashok-splunk

File: pkg/splunk/enterprise/configuration.go [hunk 6]
```diff
@@ -757,13 +768,19 @@ func getLivenessProbe(cr splcommon.MetaObject, spec *enterpriseApi.CommonSplunkS
...
+func getReadinessProbe(cr splcommon.MetaObject, instanceType InstanceType, spec *enterpriseApi.CommonSplunkSpec, additionalDelay int32) *corev1.Probe {
 	scopedLog := log.WithName("getReadinessProbe").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
 
-	readinessDelay := readinessProbeDefaultDelaySec + additionalDelay
-	if spec.ReadinessInitialDelaySeconds > readinessDelay {
-		readinessDelay = spec.ReadinessInitialDelaySeconds
+	readinessDelay := int32(readinessProbeDefaultDelaySec)
+
+	// Exclude MC, as it derives the spec from Multiple CRs.
+	if instanceType != SplunkMonitoringConsole {
```
Comment (smohan-splunk) [inline pkg/splunk/enterprise/configuration.go:777]: Add TODO for MC CR cc: @kashok-splunk

File: pkg/splunk/enterprise/util.go [hunk 7]
```diff
@@ -307,19 +322,22 @@ func ApplyAppListingConfigMap(client splcommon.ControllerClient, cr splcommon.Me
...
+	}
 	// Create App list config map
 	configMapName := GetSplunkAppsConfigMapName(cr.GetName(), crKind)
 	appListingConfigMap := splctrl.PrepareConfigMap(configMapName, cr.GetNamespace(), mapAppListing)
 
 	appListingConfigMap.SetOwnerReferences(append(appListingConfigMap.GetOwnerReferences(), splcommon.AsOwner(cr, true)))
 
-	if len(appListingConfigMap.Data) > 0 {
-		configMapDataChanged, err = splctrl.ApplyConfigMap(client, appListingConfigMap)
+	//if len(appListingConfigMap.Data) > 0 {
```
Comment (jryb) [inline pkg/splunk/enterprise/util.go:334]: So we are applying the updated ConfigMap whether its empty or not, correct?  Will this cause an extra restart if the list is emptied out?

Comment (jryb) [reply pkg/splunk/enterprise/util.go:334]: I'm thinking if we have a reconcile with app list changes, followed by a reconcile without any app changes.

**Review by jryb — APPROVED** [review]

File: pkg/splunk/enterprise/util.go (hunk 7 above)
Comment (smohan-splunk) [inline pkg/splunk/enterprise/util.go:334]: Uncomment

**Review by smohan-splunk — APPROVED** [review]

**Review by gaurav-splunk — APPROVED** [review]

File: pkg/splunk/enterprise/util.go (hunk 7 above)
Comment (sgontla) [reply pkg/splunk/enterprise/util.go:334]: Good point Jeff. I removed this check as part of the reset logic, let me restore this

**Pull request merged**